runtime/src/message_processor.rs

```diff
@@ -106,15 +133,18 @@ pub fn verify_instruction(
     };
 
     // Verify data, remember answer because comparing
-    //   a megabyte costs us multiple microseconds...
+    //  a megabyte costs us multiple microseconds...
     let mut data_changed = DataChanged::Unchecked;
     let mut data_changed = || -> bool {
         match data_changed {
-            DataChanged::Unchecked => {
-                let changed = pre.data != post.data;
-                data_changed = DataChanged::Checked(changed);
-                changed
-            }
+            DataChanged::Unchecked => match &pre.data {
+                Some(data) => {
+                    let changed = *data != post.data;
+                    data_changed = DataChanged::Checked(changed);
+                    changed
+                }
+                None => true, // Don't have pre data, assume changed
```

Comment:
how do we arrive at this match?   allocation by the program?

Reply:
This match arm is the reason for my original question to you Rob.  This arm would be hit if the logic near line 78:

`data: if *program_id != a.owner || !debitable {`

were ever out of sync with the `verify_instruction`'s use of `data_changed()`.  Ideally, I'd like the logic for both in one location, don't see a great way to do that yet.

Reply:
I get it.  cool.

This really highlights the cost of the credit only optimization, eh?  maybe we should take a hash instead of cloning the data if the data is sufficiently large?

Reply:
Yeah, I was thinking along those same lines, would need to prevent collisions since a user could pad their data to potentially cause the check to be ignored.

Reply:
you mean grind a match?

Reply:
Yeah